### ISSUE TITLE
Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.TypeBinding.hasNullTypeAnnotations()" because "providedType" is null

### ISSUE THREAD

**Issue by Meh-is-the-word** [eclipse-jdt/eclipse.jdt.core#1954]
With preview features enabled and Eclipse 24.03M1, with this code:

```
import static java.util.FormatProcessor.FMT;
import static java.lang.StringTemplate.RAW;

public final class JDK21TestingMain
{
  public static void main(final String[] args)
  {
    final int fourtyTwo = 42;
    final String str = FMT."\{fourtyTwo}";
    
    final int x=1;
    final int y=2;
    final StringTemplate st = RAW."\{x} + \{y} = \{x + y}";

    final var x1 = STR."Hello World";
    final var x2 = FMT."Hello World";
    final var x3 = RAW."Hello World";

    System.out.println(STR."Hello World");
    
    System.out.println();
  }
}
```

I get an internal compiler error:

```
Errors occurred during the build.
Errors running builder 'Java Builder' on project 'jdk21-testing'.
Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.TypeBinding.hasNullTypeAnnotations()" because "providedType" is null

Internal compiler error: java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.TypeBinding.hasNullTypeAnnotations()" because "providedType" is null at org.eclipse.jdt.internal.compiler.ast.NullAnnotationMatching.analyse(NullAnnotationMatching.java:337)	JDK21TestingMain.java
```
`package-info.java` has a `@org.eclipse.jdt.annotation.NonNullByDefault` on the package, but removing it didn't seem to make any difference.

Commenting out any reference to the new in 21 feature for StringTemplates causes the error to go away.

Had wondered if it was at all related to #1792 or #1734. Probably relevant for #544 and #1594.

**Comment by srikanth-sankaran** [comment]
Just copy pasting the code from above - on master I don't see the problem. I don't have package-info. Just the snippet from above - you anyway say package-info.java is not relevant to reproduction. I do have null analysis turned on.

**Comment by srikanth-sankaran** [comment]
Can you try with a recent I build - please mention the version string and also the exact steps to reproduce on that - Thanks.

**Comment by stephan-herrmann** [comment]
I can reproduce on what should be M1:
```
$ java -jar org.eclipse.jdt.core.compiler.batch_3.37.0.v20240104-0545.jar  --enable-preview --release 21 -warn:+nullAnnot -warn:+null -classpath <git.eclipse.jdt.core>/org.eclipse.jdt.annotation/bin/:. JDK21TestingMain.java
```

Once I got there, using current HEAD from master shows the same effect.

Btw: Too bad we don't print the stack trace ...

**Comment by Meh-is-the-word** [comment]
I just downloaded the M2 build and it also reproduces.  I will include a picture of my Null Analysis settings tab if that helps:

**Comment by stephan-herrmann** [comment]
> I just downloaded the 24.03M2 build and it also reproduces. I will include a picture of my Null Analysis settings tab if that helps. (Okay, the GitHub UI rejects my picture, so I will store it externally and link to it, I guess.) https://snipboard.io/J6PnaU.jpg

Thanks, we already found the bug and drafted a fix in #1959 - which just awaits review.

**Comment by Madjosz** [comment]
#1959 did neither fix #1734 nor #1792 so they are unrelated to this issue.

**Comment by Meh-is-the-word** [comment]
Just a note that I tried my code on the 2024.03M3 build and it now works as expected, confirming the fix of this issue.
